src/apis.md

```diff
@@ -0,0 +1,165 @@
+
+Generally speaking, types which are plain-old-data can be `Copy`. Anything
+more nuanced with any type of state shouldn't be.
+
+## Should I have `Arc` or `Rc` in my API?
+
+> it’s a code smell to have reference counts in your API design. You should hide
+> it - TM.
+
+If you must, you will need to decide between `Rc` and `Arc` - see the next
+answer for some considerations. Also, consider taking a look at the
+[`Archery` crate](https://docs.rs/archery/latest/archery/).
+
+## Should my API be thread-safe? What does that mean?
+
+In C++, a thread-safe API usually means that you can expect your API's
+consumers to use objects from multiple threads. This is difficult to make safe
+and therefore substantial extra engineering is required to make an API
+thread-safe.
+
+In Rust, things differ:
+
+* it's more normal to do things across multiple threads;
+* you don't have to worry about your callers making mistakes here because
+  the compiler won't let them;
+* you can often rely on `Send` rather than `Sync`.
+
+You certainly shouldn't be putting a `Mutex` around all your types. If your
+caller attempts to use the type from multiple threads, the compiler will
+simply stop them. It is the responsibility of the caller to use things
+safely.
+
+> if the library has `Arc` or `Rc` in the APIs, it may be making choices about
+> how you should instantiate stuff, and that’s rude. - AF.
+
+There's a reasonable chance that your API can be used in parallel threads
+by virtue of `Send` and `Sync` being automatically derived. But - you should
+think through the usage model for your API clients and ensure that's true.
+
+```rust
+use std::cell::RefCell;
+use std::collections::VecDeque;
+use std::sync::Mutex;
+use std::thread;
+
+// Absolutely no effort required to pizza, cooking or eating procedures
+// to be usable in a threaded context
+mod pizza_api {
+
+    use std::thread;
+    use std::time::Duration;
+
+    pub struct Pizza {
+        // automatically 'Send'
+        _anchovies: u32,
+        _pepperoni: u32,
+    }
+
+    pub fn make_pizza() -> Pizza {
+        println!("cooking...");
+        thread::sleep(Duration::from_millis(10));
+        Pizza {
+            _anchovies: 0, // yuck
+            _pepperoni: 32,
+        }
+    }
+
+    pub fn eat_pizza(_pizza: Pizza) {
+        println!("yum")
+    }
+}
+
+fn main() {
+    let pizza_queue = Mutex::new(RefCell::new(VecDeque::new()));
+    thread::scope(|s| {
+        s.spawn(|| {
+            let mut pizzas_eaten = 0;
+            while pizzas_eaten < 100 {
+                if let Some(pizza) = pizza_queue.lock().unwrap().borrow_mut().pop_front() {
+                    pizza_api::eat_pizza(pizza);
+                    pizzas_eaten += 1;
+                }
+            }
+        });
+        s.spawn(|| {
+            for _ in 0..100 {
+                let pizza = pizza_api::make_pizza();
+                pizza_queue.lock().unwrap().borrow_mut().push_back(pizza);
+            }
+        });
+    });
+}
```

Comment:
Be explicit about what this example code shows. E.g. is it that `Pizza` itself does not involve `Mutex` and so on, and that the caller uses those if they need to and not otherwise? If so, say that. "In this example, the `Pizza` is a simple struct. Parallel chomping of many `Pizza`s is left to the caller — `Pizza` composes in the natural way with `VecDeque`, `RefCell`, and `Mutex`, but callers with simpler needs don't need to worry about that..." or whatever.